File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/util/TypeFaceUtil.java
package com.byagowi.persiancalendar.util;

import android.content.Context;
import android.graphics.Typeface;
import android.util.Log;

import java.lang.reflect.Field;

// https://gist.github.com/artem-zinnatullin/7749076
public class TypeFaceUtil {

    /**
     * Using reflection to override default typeface
     * NOTICE: DO NOT FORGET TO SET TYPEFACE FOR APP THEME AS DEFAULT TYPEFACE WHICH WILL BE OVERRIDDEN
     *
     * @param context                    to work with assets
     * @param defaultFontNameToOverride  for example "monospace"
     * @param customFontFileNameInAssets file name of the font from assets
     */
    public static void overrideFont(Context context, String defaultFontNameToOverride, String customFontFileNameInAssets) {
        try {
            final Typeface customFontTypeface = Typeface.createFromAsset(context.getAssets(), customFontFileNameInAssets);

            final Field defaultFontTypefaceField = Typeface.class.getDeclaredField(defaultFontNameToOverride);
            defaultFontTypefaceField.setAccessible(true);
            defaultFontTypefaceField.set(null, customFontTypeface);
        } catch (Exception e) {
            Log.e("TAG", "Can not set custom font " + customFontFileNameInAssets + " instead of " + defaultFontNameToOverride);
        }
    }
}

File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/util/TypeFaceUtil.kt
package com.byagowi.persiancalendar.util

import android.content.Context
import android.graphics.Typeface
import android.util.Log

// https://gist.github.com/artem-zinnatullin/7749076
object TypeFaceUtil {

  /**
   * Using reflection to override default typeface
   * NOTICE: DO NOT FORGET TO SET TYPEFACE FOR APP THEME AS DEFAULT TYPEFACE WHICH WILL BE OVERRIDDEN
   *
   * @param context                    to work with assets
   * @param defaultFontNameToOverride  for example "monospace"
   * @param customFontFileNameInAssets file name of the font from assets
   */
  fun overrideFont(context: Context, defaultFontNameToOverride: String, customFontFileNameInAssets: String) {
    try {
      val customFontTypeface = Typeface.createFromAsset(context.assets, customFontFileNameInAssets)

      val defaultFontTypefaceField = Typeface::class.java.getDeclaredField(defaultFontNameToOverride)
      defaultFontTypefaceField.isAccessible = true
      defaultFontTypefaceField.set(null, customFontTypeface)
    } catch (e: Exception) {
      Log.e("TAG", "Can not set custom font $customFontFileNameInAssets instead of $defaultFontNameToOverride")
    }

  }
}
